keras_cv/models/segmentation/deeplab_v3/deeplab_v3.py

```diff
@@ -0,0 +1,260 @@
+            to use as image input for the model.
+
+    Examples:
+    ```python
+    import tensorflow as tf
+    import keras_cv
+
+    images = tf.ones(shape=(1, 96, 96, 3))
+    labels = tf.zeros(shape=(1, 96, 96, 1))
+    backbone = keras_cv.models.ResNet50V2Backbone(input_shape=[96, 96, 3])
+    model = keras_cv.model.segmentation.DeepLabV3(
+        num_classes=1,
+        backbone=backbone,
+        low_level_feature_layer_name="v2_stack_0_block1_2_relu"
+    )
+
+    # Evaluate model
+    model(images)
+
+    # Train model
+    model.compile(
+        optimizer="adam",
+        loss=keras.losses.BinaryCrossentropy(from_logits=False),
+        metrics=["accuracy"],
+    )
+    model.fit(images, labels, epochs=3)
+    ```
+    """
+
+    def __init__(
+        self,
+        num_classes,
+        backbone,
+        spatial_pyramid_pooling=None,
+        low_level_feature_layer_name=None,
+        projection_filters=None,
+        segmentation_head=None,
+        segmentation_head_activation="softmax",
+        dropout=0.0,
+        input_shape=(None, None, 3),
+        input_tensor=None,
+        **kwargs,
+    ):
+        if not isinstance(backbone, keras.layers.Layer) or not isinstance(
+            backbone, keras.Model
+        ):
+            raise ValueError(
+                "Argument `backbone` must be a `keras.layers.Layer` instance "
+                f" or `keras.Model`. Received instead "
+                f"backbone={backbone} (of type {type(backbone)})."
+            )
+
+        inputs = utils.parse_model_inputs(input_shape, input_tensor)
+
+        if input_shape[0] is None and input_shape[1] is None:
+            input_shape = backbone.input_shape[1:]
+            inputs = keras.Input(tensor=input_tensor, shape=input_shape)
+
+        if input_shape[0] is None and input_shape[1] is None:
+            raise ValueError(
+                "Input shapes for both the backbone and DeepLabV3 cannot be "
+                "`None`. Received: input_shape={input_shape} and "
+                "backbone.input_shape={backbone.input_shape[1:]}"
+            )
+
+        height = input_shape[0]
+        width = input_shape[1]
+
+        feature_map = backbone(inputs)
+        if spatial_pyramid_pooling is None:
+            spatial_pyramid_pooling = SpatialPyramidPooling(
+                dilation_rates=[6, 12, 18]
+            )
+
+        spp_outputs = spatial_pyramid_pooling(feature_map)
+
+        encoder_outputs = keras.layers.UpSampling2D(
+            size=(
+                height // 4 // spp_outputs.shape[1],
+                width // 4 // spp_outputs.shape[2],
+            ),
+            interpolation="bilinear",
+        )(spp_outputs)
+
+        if low_level_feature_layer_name is not None:
+            low_level_feature_extractor = keras.Model(
+                inputs=backbone.input,
+                outputs=backbone.get_layer(low_level_feature_layer_name).output,
```

Comment:
This seems like a bit of a hack -- do we expect this low level feature layer to correspond to a feature pyramid level? If so, that would be a much more elegant way to extract this.

Reply:
I did some digging, and it looks like the output from the `"v2_stack_0_block1_2_relu"` layer that you're using has shape `(None, 24, 24, 64)`.

The P2 pyramid output (which is many layers later in the model) has shape `(None, 24, 24, 256)` -- is this intentional? Can we not just use the P2 output here?

(Apologies if I'm just ignorant of the details of DeepLab)

Reply:
Hi @ianstenbit 

The DeepLabV3+ paper mentions the following regarding the low-level features

> The encoder features are first bilinearly upsampled by a factor of 4 and then concatenated with the corresponding low-level features from the network backbone that have the same spatial resolution (e.g., Conv2 before striding in ResNet-101)

I think this means that the ideal layer of the backbone from which we get the low-level features is dependent on the architecture of the backbone and is more like a hyperparameter, all that has to be made sure is that the dimensions of the low-level features has the same spatial resolution as the  encoded features.

Would love to know your thoughts on this!

Reply:
This is exactly what the `pyramid_level_inputs` in backbones are for -- to get features of a given resolution.

For example, `backbone.pyramid_level_inputs["P3"]` gets you the intermediate backbone outputs where the shape is downscaled by a factor of `2^3` from the input images ("PN" refers to `2^n` smaller)

Reply:
So probably we just want to extract the `P2` outputs

Reply:
@ianstenbit 
Thanks for the clarification!